### PR TITLE
Advanced SEO: Save custom title format to API

### PR DESCRIPTION
**Pull request by dmsnell**
<del>**DO NOT MERGE** waiting on changes to API whereas partial updates to the custom title formats currently overwrite all title formats.</del> <ins>Merged in r138837-wpcom</ins>

Part of #5834

The prior work in #6467 and #6071 added a UI element to edit custom SEO
title formats inside of Calypso but they did not save the data back
through the API to the servers.

This patch piggy-backs on the settings in `form-seo.jsx` to send changes
setting updates to WordPress.com.

New `fromApi` and `toApi` functions have been created in
`components/seo/meta-title-editor/mappings.js` to foster reusability for
the conversion code to transform between API-native and Calypso-native
formats.

A new `onChange` hook has also been added to the editor so that the SEO
form can be notified of changes. These changes are stateful at the
component level and not at the Redux state level because there remains
considerable work to bring the Redux tree to the point where it need to
be to make this smooth.

Included in these changes are some reformattings of the **lodash**
imports.

There should be **no visual changes in this PR**.

**Remaining Questions**
 - Should we be moving the `mappings.js` file into a place better suited
   for it? Maybe into `client/state/sites/` I'm not sure if that would
   be better or if its current home isn't the best place for it to be.
   It's currently only marginally referenced outside of the component.

**Testing**

You will need a site **with a business plan** and you will need to be **running a local dev environment** to test this because all public-facing changes are still hidden behind a feature flag.

Navigate to **My Sites** > **Settings** > **SEO** and notice the _Meta Title Format_ editor. Most likely it will be blank for you. If you have not selected a site with a business plan, the editor will still be there but it won't work.

You should be able to enter some combination of text and tags/chips/tokens in the boxes for each kind of page type. You may want to verify the following:

 - [x] The formats switch out appropriately if you selected different page types
 - [x] Those changes should persist when switching back to a previous page type
 - [x] If you attempt to navigate away from the page without hitting `Save Settings` then it should alert you of unsaved changes in a popup.
 - [x] If you hit `Save Settings` you should find it successfully send of a network request to the site settings endpoint.
 - [ ] If you reload the page after saving, your changes should come back or persist.

**Notes**
 - I'm not sure how the unsaved-changes popup works. Literally, I didn't even try to make it work right and magically it just did. Likely I have been lucky but it will eventually break. @roundhill maybe you have a quick word of advice.
 - There's a race-condition after saving changes that exists because we aren't updating the Redux state tree directly on title format saves. Instead, once we update the settings on the server, we request an update to the site settings via `site.fetchSettings()`. If we reload the page after the title format save but before those settings are fetched again, it will appear like we have lost the changes. However, if we reload the page again, we are likely to find that they reappear. I'm not sure what the best way to handle this is. I think the probability of experiencing this bug is low due to the likely use-flow of the editor.

cc: @rodrigoi @vindl @gwwar @mtias

Test live: https://calypso.live/?branch=update/seo-title/save-to-api

### CODE REVIEW

**Comment by gwwar** [comment]
> I'm not sure how the unsaved-changes popup works.

It's probably the protectForm mixin that adds an onbeforeunload handler.
https://github.com/Automattic/wp-calypso/blob/master/client/my-sites/site-settings/form-seo.jsx#L87

**Comment by lamosty** [comment]
Nice work!

All the things mentioned in testing instructions are working. However, I found a bug with saving the title formats (maybe the thing you described in the notes section). [See Video](https://drive.google.com/file/d/0B8e-IpRwSdjtNzFCclhwVHZMT1U/view). Basically, when I add something to the meta title format, save and reload, it will appear. But, when I add another thing to the same tab (e.g. Front Page), save and reload, it disappears. I have to add it again, hit save and only then it persists after reload. The same for item deletion. It's a bit annoying.

File: client/components/seo/meta-title-editor/mappings.js [hunk 1]
```diff
@@ -50,3 +57,13 @@ const tokenToTag = n =>
...
+
+export const toApi = compose(
+	partialRight( mapKeys, rearg( snakeCase, 1 ) ), // 1 -> key from ( value, key )
+	partialRight( mapValues, nativeToRaw )          // native objects to raw strings
+);
+
+export const fromApi = compose(
+	partialRight( mapKeys, rearg( camelCase, 1 ) ), // 1 -> key from ( value, key )
+	partialRight( mapValues, rawToNative )          // raw strings to native objects
+);
```
Comment (lamosty) [inline client/components/seo/meta-title-editor/mappings.js]: Pretty nice functional? programming here. However, to be honest, it's hard to follow if you don't know all the functions used here. I had to read through lodash doc and it took me quite some time to finally understand this. I'm not sure if we are not sacrificing readability/time spent to review and understand this for code beauty/quality.

Comment (dmsnell) [inline client/components/seo/meta-title-editor/mappings.js]: @lamosty I sympathize with the fact that many of these functions aren't familiar to many of us, but on the other hand, we have very much adopted **lodash** into the project as a standard library and I don't want to shy away from it just because some of it might be new.

The biggest critique I could agree with is the part about `rearg( …, 1 )`, which was going to be bad no matter how I sliced it. This is due to the way that `mapKeys` works, by sending `( value, key|index, collection )` into the callback. I'm personally opposed to creating variables that we don't end up using because of the confusion it brings in: why is that variable there? why is it named that way? did someone mess up by removing it from the body before but leaving it in? Anyway, even with a closure here like `(value, key) => camelCase( key )` I think the same knowledge about `mapKeys` is needed.

Additionally, `mapKeys`, `mapValues`, and `rearg` are all pretty simple concepts wherein they don't individually do all that much. Hopefully they will be easy to remember and maybe you will find more uses for them in the future.

P.S. The combination of **Dash** and **Alfred** make for an amazing documentation experience. If you haven't looked into them before, it may be worth installing them. I'm only ever a few keystrokes away from documentation on _lodash_, _JavaScript_, _CSS_, _React_, etc…

P.P.S. I added an explanatory comment in f8e681d for why those methods are there.

**Comment by dmsnell** [comment]
> when I add another thing to the same tab (e.g. Front Page), save and reload, it disappears

good spotting. this is indeed related to the same race condition mentioned in the notes, but I believe we should be able to prevent this one. we designed the API to accept a partial list of mappings, so if we iterate on it, we should be able to only send the changes which would leave the other tabs untouched.

File: client/components/seo/meta-title-editor/mappings.js (hunk 1 above)
Comment (lamosty) [inline client/components/seo/meta-title-editor/mappings.js]: Nice explanation @dmsnell.

I always learn something new when reviewing your PRs so maybe it's great you are doing things in this way. :)

Hmm, didn't know about Dash/Alfred. I have Ubuntu, so will have to find some alternatives. But nice idea, thanks!

**Comment by dmsnell** [comment]
> bug with saving the title formats

this bug should be mostly eliminated with the changes to the API mentioned in the PR description. there is still a race condition, but now it only exists on a specific `type` for the formats, that is, it won't mess up or clear out other types in the reload cycle.

**Comment by rralian** [comment]
The code looks good and clean and makes sense to me. I checked off all the things that are working correctly for me. But unfortunately I'm running into a couple of issues when testing this.
- The Meta Title Format works the same for me for a free site versus a business site.
- I get a javascript error when I try to save (for both free and business).
- The setting is not actually saved... it doesn't appear when I reload. Tried several times.

javascript error

Sounds like a reduxy issue to me, and I ran into this with something I was working on before, but I'm not really sure what's causing it.

Some UX thoughts that came to mind but may not be pertinent to this PR but I'll share anyway; How are we going to communicate that the form is disabled for non-business sites? How will we communicate that this is a business-only feature and that they can upgrade to activate it?

**Comment by dmsnell** [comment]
> I get a javascript error when I try to save (for both free and business).

@rralian fixed in 849ac2c

**Comment by dmsnell** [comment]
> The setting is not actually saved... it doesn't appear when I reload. Tried several times.

@rralian were you trying to save to a business site? it definitely won't save on free or premium sites.

**Comment by dmsnell** [comment]
> The Meta Title Format works the same for me for a free site versus a business site.

I'm not certain what our plan is. I think that the last time I looked, we were disabling it by graying it out. I'm cool with doing whatever we feel is best.

**Comment by rralian** [comment]
> @rralian were you trying to save to a business site? it definitely won't save on free or premium sites.

Yes, I tried saving with a free site and a business site and ran into the problem with both sites. I see you pushed a fix for the javascript error. Let me know when you want me to test again.

**Comment by rralian** [comment]
I tested again with a business site and had the same problems. The issue is that after I save a new string, and I reload the page, I don't see that saved string. I just see whatever setting was there before I saved. Sometimes I can see the new string after a _second_ reload, but sometimes it takes several reloads before I see the new string. I've tried this several times and I can recreate this consistently. Here's a video. https://cloudup.com/cbWXZefPK2C

**Comment by dmsnell** [comment]
> after I save a new string, and I reload the page, I don't see that saved string

yes, I see now. You have stumbled upon the race condition that we knew exists and would exist since we started working on this. We could spend some time trying to fix this (I'm pretty sure it's related to that big SitesList) but I don't feel like it's likely to be a big bug. Once the site information gets updated in Calypso again the data should be back in sync. I stopped working on this route because of the scale of the effort I was figuring would be required

**Comment by rralian** [comment]
> We could spend some time trying to fix this (I'm pretty sure it's related to that big SitesList) but I don't feel like it's likely to be a big bug.

I looked a little further and I think the issue is that you're reading the `advanced_seo_title_formats` only from the `/me/sites` endpoint. I'm surprised it was added directly to the site object rather than being requested from the `/site/settings` object, because I wouldn't think this is very frequently needed, and it's probably best not to slow down the `/me/sites` endpoint by adding this  lookup to every site. The site-settings code requests fresh data from the `/site/settings` endpoint via the `site.fetchSettings()` call, so we don't need to wait for the much slower 30-second polling on `/me/sites` to get fresh data. Not saying it's a great design, but that's why settings are saved more consistently in "general" than for these new meta settings.

I don't think we need to address this in this PR which is more about making sure we are saving data in the right format, so 👍  on that. But I do think we should address this before we would launch this to production. It doesn't need to be addressed the same way the other settings pages work, but I think you should be able to reload the page and see your saved settings more reliably.
